qutebrowser/misc/sql.py

```diff
@@ -161,7 +161,8 @@ def __init__(self, name, fields, constraints=None, parent=None):
         self._name = name
 
         constraints = constraints or {}
-        column_defs = ['{} {}'.format(field, constraints.get(field, ''))
+        default = 'NOT NULL'
+        column_defs = ['{} {}'.format(field, constraints.get(field, default))
```

Comment:
The `default` variable should be `default_constraint` for clarity, SQL has `DEFAULT` values which might be set as `constraints={"foo": "DEFAULT 1"}`.

Also note that you are still not enforcing the `NOT NULL` constraint. For example:

- If you had `constraints={"foo": "UNIQUE"}` then `NULL` values are still allowed in the `foo` column: see https://sqlite.org/faq.html#q26.
- If you had `constraints={"foo": "CHECK foo > 42"}`, then `NULL` is still allowed in `foo`.
- If you had `constraints={"foo": "FOREIGN KEY (foo) REFERENCES some_table(foo)"}` then the `foo` column would still be nullable.

I guess this is a price for having a custom objective wrapper around SQL... Sometimes I wonder why we don't use [SqlAlchemy](http://www.sqlalchemy.org/) which nicely handles all of these issues &ndash; the only argument that comes to mind is the interoperability with Qt, but maybe it's time to explore it? [Here](https://stackoverflow.com/a/12206759/4180822) is a proof that it was at least possible at some point, though the project seems to be dead now.

Reply:
We could just add these to `CompletionHistory.__init__` rather than having a magic default constraint:


```
def __init__(self, parent=None):
    super().__init__("CompletionHistory", ['url', 'title', 'last_atime'],
                     constraints={'url': 'PRIMARY KEY NOT NULL'
                                  'title': 'NOT NULL'
                                  'atime': 'NOT NULL'
                                  'redirect': 'NOT NULL'},
                     parent=parent)
```

SqlAlchemy might be a good idea though.